src/allmydata/grid_manager.py

```diff
@@ -0,0 +1,348 @@
+import sys
+import json
+from datetime import (
+    datetime,
+    timedelta,
+)
+
+from allmydata.crypto import (
+    ed25519,
+)
+from allmydata.util import (
+    fileutil,
+    base32,
+)
+
+
+class _GridManagerStorageServer(object):
+    """
+    A Grid Manager's notion of a storage server
+    """
+
+    def __init__(self, name, public_key, certificates):
+        self.name = name
+        self._public_key = public_key
+        self._certificates = [] if certificates is None else certificates
+
+    def add_certificate(self, certificate):
+        self._certificates.append(certificate)
+
+    def public_key(self):
+        return ed25519.string_from_verifying_key(self._public_key)
+
+    def marshal(self):
+        return {
+            u"public_key": self.public_key(),
+        }
+
+
+def create_grid_manager():
+    """
+    Create a new Grid Manager with a fresh keypair
+    """
+    private_key, public_key = ed25519.create_signing_keypair()
+    return _GridManager(
+        ed25519.string_from_signing_key(private_key),
+        {},
+    )
+
+
+def load_grid_manager(config_path, config_location):
+    """
+    Load a Grid Manager from existing configuration.
+
+    :param FilePath config_path: the configuration location (or None for
+        stdin)
+
+    :param str config_location: a string describing the config's location
+
+    :returns: a GridManager instance
+    """
+    if config_path is None:
+        config_file = sys.stdin
+    else:
+        try:
+            config_file = config_path.child("config.json").open("r")
+        except IOError:
+            raise ValueError(
+                "'{}' is not a Grid Manager config-directory".format(config_path)
+            )
+    with config_file:
+        config = json.load(config_file)
+
+    gm_version = config.get(u'grid_manager_config_version', None)
+    if gm_version != 0:
+        raise ValueError(
+            "Missing or unknown version '{}' of Grid Manager config".format(
+                gm_version
+            )
+        )
+    if 'private_key' not in config:
+        raise ValueError(
+            "Grid Manager config from '{}' requires a 'private_key'".format(
+                config_location,
+            )
+        )
+
+    private_key_bytes = config['private_key'].encode('ascii')
+    try:
+        private_key, public_key = ed25519.signing_keypair_from_string(private_key_bytes)
+    except Exception as e:
+        raise ValueError(
+            "Invalid Grid Manager private_key: {}".format(e)
+        )
+
+    storage_servers = dict()
+    for name, srv_config in config.get(u'storage_servers', {}).items():
+        if not 'public_key' in srv_config:
```

Comment:
`'public_key' not in srv_config` always seemed to read better to me and leaves less ambiguity about whether you're testing `(not 'public_key') in srv_config` or not `('public_key' in srv_config)`.

Reply:
Ah right thanks (also, linter usually catches this .. *looks* like codechecks should be running flake8 :woman_shrugging: )